packages/server/src/games.ts

```diff
@@ -58,3 +58,99 @@ export function updateGameState(io: IOServer, store: Store, roomName: string, ga
     io.in(roomName).emit(GameEvent.CurrentState, updatedGameState);
   }
 }
+
+/**
+ * Calculate the immediate new game state, as a result of a player's action. This is not necessarily the final state
+ * but an intermediate one to help the following calculations.
+ */
+function calculatePlayerAction(oldState: GameState, playerAction: PlayerAction) {
+  const activePlayer = oldState.players.filter((playerState) => playerState.username === playerAction.playerName)[0];
+  // TODO: Later on we will need to add order of players, built into the model
+  const opponent = oldState.players.filter((playerState) => playerState.username !== playerAction.playerName)[0];
+  const hand = activePlayer.hand.filter((c) => cardKey(c) !== playerAction.card);
+  const table = [...oldState.table, fromCardKey(playerAction.card)];
+
+  const newState = {
+    ...oldState,
+    activePlayer: opponent.username,
+    players: [
+      opponent,
+      {
+        ...activePlayer,
+        hand,
+      },
+    ],
+    table,
+  };
+
+  return newState;
+}
+
+/**
+ * Calculate the new game state, as it should be at the end of a player's turn.
+ */
+function calculatePlayerTurn(oldState: GameState) {
+  const { players, deck, table, latestCaptured, ...rest } = oldState;
+
+  const isRoundFinshed = oldState.players.every((player) => player.hand.length === 0);
+  const isMatchFinished = isRoundFinshed && oldState.deck.length === 0;
+  let newState = <GameState>{};
```

Comment:
same here since we are mutating anyway, you can make a const here and intialize with the old state, either making a copy or just referencing the same thing then later on you just update it's keys and you won't need to handle ...rest and the things that you are not actually updating

Reply:
Hm... Ok. Personally, for this case, I prefer copying here initially, and mutating later. So I will do that.

Reply:
ok just be aware that spread operator do shallow clone so if you have a nested object or array and mutate that part you are still mutating the initial one. Not doing here very likely but just in case. for player for example it can happen

Reply:
Oh, didn't know that. Ok, I will use the loadash one then.